### PR TITLE
add xdp support

### LINKED ISSUE #218: [FEATURE] Support more attaching types for program
Currently, we only support eBPF programs that can be auto-attached and very few other attach types (tc only). We need to implement more attach types.

### PR DESCRIPTION
**Pull request by CH3CHOHCH3**
# add xdp support

Fixes #218 

## Description

Add xdp type support for bpf-loader-lib by APIs of [libbpf_sys](https://docs.rs/libbpf-sys/latest/libbpf_sys/):

- [bpf_xdp_attach](https://docs.rs/libbpf-sys/latest/libbpf_sys/fn.bpf_xdp_attach.html)
- [bpf_xdp_detach](https://docs.rs/libbpf-sys/latest/libbpf_sys/fn.bpf_xdp_detach.html)

According to the argument list of [bpf_xdp_attach](https://docs.rs/libbpf-sys/latest/libbpf_sys/fn.bpf_xdp_attach.html) and the definition of [bpf_xdp_attach_opts](https://docs.rs/libbpf-sys/latest/libbpf_sys/struct.bpf_xdp_attach_opts.html), we can use

```C
/// @ifindex 1
/// @flags 0
/// @xdpopts {"old_prog_fd":0}
```
to pass parameters to bpf-loader-lib and the latter will use libbpf_sys to implement attach. C BPF code if here for test:

```C
#include "vmlinux.h"
#include <bpf/bpf_helpers.h>

/// @ifindex 1
/// @flags 0
/// @xdpopts {"old_prog_fd":0}
SEC("xdp")
int xdp_pass(struct xdp_md* ctx) {
    void* data = (void*)(long)ctx->data;
    void* data_end = (void*)(long)ctx->data_end;
    int pkt_sz = data_end - data;

    bpf_printk("packet size: %d", pkt_sz);
    return XDP_PASS;
}

char __license[] SEC("license") = "GPL";
```

## Test

```console
cargo test test_xdp_attach_and_detach
```

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by yunwei37 — CHANGES_REQUESTED** [review]
please add a readme.md for the xdp example, include how to run it and the results.

**Review by Officeyutong — CHANGES_REQUESTED** [review]
Seems that the CI fails to run on your example. You should fix that

**Review by Officeyutong — CHANGES_REQUESTED** [review]
Please add tests on attaching & detaching xdp programs

File: bpf-loader-rs/bpf-loader-lib/src/skeleton/preload/attach.rs [hunk 1]
```diff
@@ -30,10 +32,36 @@ impl Drop for AttachLink {
...
+                let err = unsafe { bpf_xdp_detach(*ifindex, *flags, &**opts) };
+                if err != 0 {
+                    error!("Failed to detach xdp hook");
+                }
+            }
         }
     }
 }
 
+pub(crate) fn attach_xdp(program: &Program, meta: &ProgMeta) -> Result<AttachLink> {
```
Comment (Officeyutong) [inline bpf-loader-rs/bpf-loader-lib/src/skeleton/preload/attach.rs:45]: libbpf-rs provides APIs to attach XDP. https://docs.rs/libbpf-rs/latest/libbpf_rs/struct.Program.html#method.attach_xdp . is this suitable for us?

File: bpf-loader-rs/bpf-loader-lib/src/skeleton/preload/attach.rs [hunk 2]
```diff
@@ -30,10 +32,36 @@ impl Drop for AttachLink {
...
+
+    let ifindex = xdp_extra_meta.ifindex;
+    let flags = xdp_extra_meta.flags;
+    let prog_fd = program.fd();
+
+    let mut xdp_attach_opts = Box::new(unsafe { std::mem::zeroed::<bpf_xdp_attach_opts>() });
+    xdp_attach_opts.sz = std::mem::size_of::<bpf_xdp_attach_opts>() as _;
+    xdp_attach_opts.old_prog_fd = xdp_extra_meta.xdpopts.old_prog_fd;
+
+    let err = unsafe { bpf_xdp_attach(ifindex, prog_fd, flags, &*xdp_attach_opts) };
```
Comment (Officeyutong) [inline bpf-loader-rs/bpf-loader-lib/src/skeleton/preload/attach.rs:57]: Unsafe blocks should be described with a "//SAFETY:" comment to provide a brief description on what this unsafe call does and why this unsafe call is "safe"

Comment (Officeyutong) [reply bpf-loader-rs/bpf-loader-lib/src/skeleton/preload/attach.rs:57]: See other unsafe calls for reference

**Review by Officeyutong — CHANGES_REQUESTED** [review]

File: bpf-loader-rs/bpf-loader-lib/src/skeleton/preload/attach.rs (hunk 1 above)
Comment (Officeyutong) [reply bpf-loader-rs/bpf-loader-lib/src/skeleton/preload/attach.rs:45]: It's not suitable

**Review by Officeyutong — APPROVED** [review]

**Pull request merged**